### PR TITLE
Allow subdomains in canonical domains

### PR DESCRIPTION
**Pull request by roosnic1**

### CODE REVIEW

**Comment by matthiask** [comment]
Two points:

- I think it would be better to add a list of allowed hostnames and not just subdomains...
- ... but maybe you should just use Django's `SecurityMiddleware` with `SECURE_SSL_HOST` instead? It seems to me that allowing several domains isn't a good match for a canonical domain module. But please convince me if you disagree :)

**Comment by roosnic1** [comment]
Sorry for the late response. I think it would be very useful to allow certain subdomains. In our case we want users to be able to use `mitmachen.tsri.ch` and `mitmachen.tsüri.ch` and still have them all at `mitmachen.tsri.ch`.

And if we use full hostnames instead of subdomains it will be more difficult to match them. What do you think?

**Comment by matthiask** [comment]
It still strikes me as strange that you want to have more than one canonical domain :-)

(Since you're using feincms3-sites wouldn't it work better to replace `canonical_domain.middleware.CanonicalDomainMiddleware` with `feincms3_sites.middleware.redirect_to_site_middleware` and `django.middleware.security.SecurityMiddleware`? If course you'd have to reorder the middleware a bit.)

**Comment by roosnic1** [comment]
I could do that but then i would lose the canonical domain feature. Which i still want.

**Comment by PetrDlouhy** [comment]
@matthiask @roosnic1 I would like to force the `www` subdomain on my website to all subdomains except `api` (on which I would use different DNS/proxy settings).
Is that a good use-case for PR like this? If yes, would it be interest with help to finish this?

**Comment by matthiask** [comment]
Hi @PetrDlouhy 

I'm not sure. You probably have a middleware or something which checks the domain and adds a different `request.urlconf` anyway, so that you can use a completely different URL structure for your API subdomain, or not? If yes, you'd already have a place for adding these behaviors?

My point from above still stands; I'm not sure if it makes sense to have several canonical domains; if you need more complex behaviors you are either adding your own routing middleware anyway and you can add the http/https redirect there as well. And if having multiple canonical domains makes sense, I still wouldn't want to only define a list of allowed subdomains but instead define a list of full domains.

Long story short: I still don't quite see the requested feature as a good fit for django-canonical-domain, but I'm definitely willing to work on this with you if you want to prepare a PR.

**Comment by PetrDlouhy** [comment]
@matthiask To pain whole picture of what I am trying to achieve now: We have API running on `www.blenderkit.com/api/v1/` right now. We are running through CloudFlare with their proxy features which is beneficial on most URLs.
We have problem connecting from our add-on to API (surprisingly browser connections are OK) in some countries (Nigeria, Kazakhstan). Our suspicion is, that these countries are blocking traffic to certain sites based on IP addresses and that if we are using CloudFlare proxy, those IP addresses are shared with those certain sites.

Right now I don't want to change anything except adding a copy of API URLs on `api.blenderkit.com/api/v1/` with unproxied routing. So I could just add an exception for `api` subdomain to `django-canonical-domain` would be ideal in this situation.

I am not particularly in favor of either variant. Both list of subdomains or list of complete domains would work for me. There is also third variant - list of matching patterns.

Also, I would not call this feature having multiple canonical domains. I am thinking about it more like there is one canonical domain and then several exceptions.
While I understand the desire to keep this app as simple and not sink into complex usage scenarios, I think this feature could be very simple and enable usage of the package in many scenarios with slightly more requirements than the basic behavior.

**Comment by matthiask** [comment]
Thanks for the thorough explanation!

> Also, I would not call this feature having multiple canonical domains. I am thinking about it more like there is one canonical domain and then several exceptions.

This is a very convincing way to look at it, and makes a lot of sense to me. 

> I am not particularly in favor of either variant. Both list of subdomains or list of complete domains would work for me. There is also third variant - list of matching patterns.

I like regular expressions a lot, but many people find them hard to use. I think a list of complete domains is the way to go.

**Comment by PetrDlouhy** [comment]
@matthiask OK. Thanks, I will try to look at this and update the code in following days.

**Comment by matthiask** [comment]
Replaced by #3

**Comment by PetrDlouhy** [comment]
For anyone reading this in the future: According to our tests it DOES help turning off the CloudFlares proxy to overcome the network blocking in such countries.
